torchinductor/codegen/triton.py

```diff
@@ -955,129 +961,136 @@ def __init__(self, scheduler):
+                if len(tiling1) > 2:
+                    if len(tiling2) > 2:
+                        return tiling1 == tiling2 == tiling3
+                    else:
+                        return tiling1 == tiling3
+                elif len(tiling2) > 2:
+                    return tiling2 == tiling3
+
+            return True
+
+        if not node1.is_reduction() and node2.is_reduction():
+            assert rnumel1 == 1 and rnumel2 != 1
+            if numel1 == numel2 * rnumel2:
+                if not all(
+                    TritonKernel.is_compatible((numel2, rnumel2), n.get_ranges())
+                    for n in node1.get_nodes()
                 ):
-                    node.mark_run()
-                    node_schedule.append(node)
-                    node.mark_fusable()
-                else:
-                    log.debug(
-                        "rescheduling due to not is_compatible(%s, %s)",
-                        (numel, reduction_numel),
-                        node.get_ranges(),
+                    return False
+                if config.triton.tiling_prevents_reduction_fusion:
+                    return self.select_tiling(node1.get_nodes(), numel1) in (
+                        (numel1, 1),
+                        (numel2, rnumel2, 1),
                     )
-                    nodes_to_reschedule.append(node)
+                return True
 
-            # scheduler.pop_group will keep iterating all reachable fusable nodes
-            reductions_to_mark_fusable = []
-            for node in self.scheduler.pop_group((numel, reduction_numel)):
-                node.mark_run()
-                node_schedule.append(node)
-                reductions_to_mark_fusable.append(node)
-            # mark reductions fusable later as they rely on the loop break below
-            for node in reductions_to_mark_fusable:
-                node.mark_fusable(broadcast_after_reduce=True)
-
-            node_schedule.append(DisableReduction)  # close reduction loop
-            # Add more pointwise with fewer dimensions
-            for node in self.scheduler.pop_group((numel, sympy.Integer(1))):
-                node.mark_run()
-                node_schedule.append(node)
-                node.mark_fusable()
-            node_schedule.append(EnableReduction)  # open new reduction loop
+            return numel1 == numel2
 
-            if self.is_better_tiling_ready(numel, reduction_numel):
-                # early exit to prevent a fusion that would result in worse tiling
-                break
+        assert node1.is_reduction() and not node2.is_reduction()
+        # swap args to hit the case above
+        return self.can_fuse_horizontal(node2, node1)
 
-        self.scheduler.enqueue(nodes_to_reschedule)
-        return node_schedule
+    can_fuse_vertical = can_fuse
+    can_fuse_horizontal = can_fuse
 
-    def codegen(self, numel, reduction_numel):
+    def codegen_nodes(self, nodes):
```

Comment:
This code seems like it depends heavily on the order of `nodes`. What kind of guarantees do we have on that from the prefusion pass? And are there multiple valid orders that could lead to different performance?

Reply:
Yes, we depend on fusion pass to construct a legal order of nodes.

Order matters, as seen in #938